File: guides to algorithms/Guia06/Java/HashTable/HashTable.java
package hashtable;

public class HashTable {

    ListaLigada[] T = null;
    int m = 0;
    int n = 0;
    
    private int hash_function_division(int k){
        return k % this.m;
    }
    
    private int hash_function_mult(int k){
        final double A = (Math.sqrt(5) - 1) / 2.0;
        return (int) Math.floor(this.m * ((k * A) % 1));
    }
    
    public HashTable(int m, int n){
        this.m = m;
        this.n = n;
        this.T = new ListaLigada[m];
        for(int k = 0; k < m; k++){
            this.T[k] = new ListaLigada(n);
        }
    }
    
    public void insert(int key){
        int h = this.hash_function_mult(key);
        System.out.println("chave = " + key + " hash = " + h);
        this.T[h].insert(key);
    }
    
    public String search(int k){
        int h = this.hash_function_mult(k);
        int x = this.T[h].search(k);
        StringBuilder res = new StringBuilder();
        res.append("----\nh = ");
        res.append(h);
        res.append(" x = ");
        res.append(x);
        return res.toString();
    }
    
    @Override public String toString(){
        String s = "";
        int k = 0;
        for(ListaLigada l : this.T){
            s += "\nLista = " + k;
            s += l;
            k++;
        }
        return s;
    }        
}

File: Guias_praticos/Guia06/Python/HashTable/HashTable.py
from io import StringIO
import math
from ListaLigada import ListaLigada

class HashTable:

    m = 0
    n = 0

    def __init__(self, m, n):
        self.m = m
        self.n = n
        self.T = ListaLigada(m)
        for k in range(0, m):
            self.T[k] = ListaLigada(n)

    def hash_function_division(self, k):
        return k % self.m

    def hash_function_mult(self, k):
        A = (math.sqrt(5) - 1) / 2
        return math.floor(self.m * ((k * A) % 1))

    def insert(self, key):
        h = self.hash_function_mult(key)
        #print("chave = {} hash = {}".format(key, h))
        self.T[h].insert(key)

    def search(self, k):
        h = self.hash_function_mult(k)
        x = self.T[h].search(k)        
        res = StringIO()
        res.write("----\nh = ");
        res.write("{}".format(h));
        res.write(" x = ");
        res.write("{}".format(x));
        return res.getvalue()

    def toString(self):
        s = ""
        k = 0
        for x in range(0, self.m):
            l = self.T[x]
            s += "\nLista = {}\n".format(k)
            s += l.toString()
            k += 1
        return s
